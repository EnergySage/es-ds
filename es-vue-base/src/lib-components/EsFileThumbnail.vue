<template>
    <div>
        <!-- MOBILE PREVIEW START -->
        <div
            class="mobile-preview"
            :class="mobileView ? '' : 'd-block d-md-none'">
            <div
                class="card thumbnail-border">
                <b-row
                    align-v="center"
                    class="p-3">
                    <b-col
                        cols="2">
                        <div
                            v-if="!loading"
                            class="text-success p-0">
                            <icon-circle-check />
                        </div>
                        <div v-else>
                            <es-progress-circle
                                :value="percentLoaded"
                                :show-percentage="false"
                                height="24px" />
                        </div>
                    </b-col>
                    <b-col
                        cols="8"
                        class="p-0">
                        <div class="font-weight-bolder text-gray-800 pl-0">
                            {{ fileName }}
                        </div>
                    </b-col>
                    <b-col
                        cols="2">
                        <b-link
                            class="text-gray-900 float-right icon-button"
                            aria-label="close-file-mobile"
                            @click="$emit('removeFile',fileName)">
                            <icon-trash-can />
                        </b-link>
                    </b-col>
                </b-row>
                <b-row
                    align-v="center"
                    class="px-3 pb-3">
                    <b-col cols="2" />
                    <b-col
                        cols="10"
                        class="pl-0">
                        <template v-if="mimeType && mimeType.includes('image') && fileSource">
                            <b-link
                                v-if="!loading"
                                aria-label="show-preview-mobile-image"
                                @click="$emit('showPreview',fileName)">
                                {{ previewText }}
                            </b-link>
                        </template>
                        <template v-else>
                            <b-link
                                v-if="!loading"
                                aria-label="show-preview-mobile-file"
                                :href="fileSource"
                                target="_blank">
                                {{ previewText }}
                            </b-link>
                        </template>
                    </b-col>
                </b-row>
            </div>
        </div>
        <!-- MOBILE PREVIEW END -->
        <!-- DESKTOP/TABLET PREVIEW START -->
        <div
            class="desktop-preview"
            :class="mobileView ? 'd-none' : 'd-none d-md-block'">
            <div class="thumbnail-outer-wrapper position-relative">
                <div class="float-right d-flex">
                    <b-link
                        v-if="!loading"
                        aria-label="close-file-desktop"
                        class="text-gray-800 text-decoration-none icon-button"
                        @click="$emit('removeFile',fileName)">
                        <icon-circle-x />
                        <div class="svg-fill-wrapper bg-white" />
                    </b-link>
                </div>

                <b-link
                    v-if="mimeType.includes('image')"
                    class="text-decoration-none text-gray-800"
                    aria-label="show-preview-desktop-image"
                    :disabled="loading"
                    @click="$emit('showPreview',fileName)">
                    <div
                        :class="{
                            'card': true,
                            'overflow-hidden': !loading,
                            'thumbnail-inner-wrapper-x': !loading,
                            'thumbnail-inner-wrapper': loading,
                        }">
                        <template v-if="!loading">
                            <div class="h-100">
                                <b-img
                                    class="image-preview"
                                    :src="fileSource"
                                    :alt="fileName" />
                            </div>
                        </template>
                        <template v-else>
                            <div class="h-100 d-flex align-items-center justify-content-center">
                                <es-progress-circle
                                    v-if="percentLoaded"
                                    :show-percentage="percentLoaded"
                                    :value="percentLoaded"
                                    circle
                                    height="65px" />
                                <b-spinner
                                    v-if="!percentLoaded"
                                    role="status"
                                    label="Loading" />
                            </div>
                        </template>
                    </div>
                </b-link>
                <b-link
                    v-if="mimeType.includes('application')"
                    class="text-decoration-none text-gray-800"
                    aria-label="show-preview-desktop-file"
                    :disabled="loading"
                    :href="fileSource"
                    target="_blank">
                    <div
                        :class="{
                            'card': true,
                            'overflow-hidden': !loading,
                            'thumbnail-inner-wrapper-x': !loading,
                            'thumbnail-inner-wrapper': loading,
                        }">
                        <template v-if="!loading">
                            <div
                                class="h-100 d-flex align-items-center justify-content-center">
                                <icon-file-pdf v-if="mimeType.includes('pdf')" />
                                <icon-file-docx v-else-if="mimeType.includes('docx')" />
                                <icon-file-doc v-else />
                            </div>
                        </template>
                        <template v-else>
                            <div class="h-100 d-flex align-items-center justify-content-center">
                                <es-progress-circle
                                    :value="percentLoaded"
                                    height="65px" />
                            </div>
                        </template>
                    </div>
                </b-link>
            </div>
            <div class="font-weight-bolder text-truncate pt-2 pt-md-3">
                {{ fileName }}
            </div>
            <div
                v-if="fileSize"
                class="font-weight-normal">
                {{ fileSize }}
            </div>
        </div>
        <!-- DESKTOP/TABLET PREVIEW END -->
    </div>
</template>

<script>
import { BImg, BLink } from 'bootstrap-vue';
import IconCircleCheck from '../lib-icons/icon-circle-check.vue';
import IconCircleX from '../lib-icons/icon-circle-x.vue';
import IconFileDoc from '../lib-icons/icon-file-doc.vue';
import IconFileDocx from '../lib-icons/icon-file-docx.vue';
import IconFilePdf from '../lib-icons/icon-file-pdf.vue';
import IconTrashCan from '../lib-icons/icon-trash-can.vue';
import EsProgressCircle from './EsProgressCircle.vue';

export default {
    name: 'EsFileThumbnail',
    components: {
        BImg,
        BLink,
        IconCircleCheck,
        IconCircleX,
        IconTrashCan,
        IconFilePdf,
        IconFileDocx,
        IconFileDoc,
        EsProgressCircle,
    },
    props: {
        /**
         * File Name
         * Already formatted, e.g. 'bill.png'
         *
         */
        fileName: {
            type: String,
            default: 'test.png',
            required: false,
        },
        /**
         * Mime Type
         * File format, e.g. 'application/pdf', 'image/jpg'
         */
        mimeType: {
            type: String,
            default: 'application/pdf',
            required: false,
        },
        /**
         * File Source
         * If applicable, the source of the image
         *
         */
        fileSource: {
            type: String,
            default: '',
            required: false,
        },
        /**
         * File Size
         * Already formatted, e.g. '2.4 MB'
         *
         */
        fileSize: {
            type: String,
            default: '',
            required: false,
        },
        /**
         * Text on mobile to preview the file
         *
         */
        previewText: {
            type: String,
            default: 'View preview',
            required: false,
        },
        /**
         * The progress of the upload, out of 100
         */
        percentLoaded: {
            type: Number,
            default: 0,
        },
        /**
         * To show the mobile view in desktop
         */
        mobileView: {
            type: Boolean,
            default: false,
        },
    },
    computed: {
        loading() {
            return (this.percentLoaded !== 100);
        },
    },
};
</script>

<style lang="scss" scoped>
@import '~@energysage/es-bs-base/scss/includes';

.desktop-preview {
    width: 165px;
}

.thumbnail-border {
    border: 2px solid $card-border-color;
}

.thumbnail-border-failure {
    border: 2px solid $danger;
}

.thumbnail-outer-wrapper {
    height: 165px;
    width: 165px;
    z-index: 2;
    @include media-breakpoint-down(md) {
        height: 110px;
        width: 100px;
    }
}

.thumbnail-inner-wrapper-x {
    border: 2px solid $card-border-color;
    height: 160px;
    top: -18px;
    width: 160px;
    z-index: -1;

    @include media-breakpoint-down(md) {
        height: 105px;
        width: 95px;
    }

    &:hover {
        border: 2px solid $gray-900;
    }
}

.thumbnail-inner-wrapper {
    border: 2px solid $card-border-color;
    height: 160px;
    overflow: hidden;
    top: 8px;
    width: 160px;
    z-index: -1;

    @include media-breakpoint-down(md) {
        height: 105px;
        width: 95px;
    }
}

.icon-button {
    position: relative;
    z-index: 2;
}

a:hover {
    color: $gray-800;
}

.svg-fill-wrapper {
    border-radius: 1rem;
    height: 20px;
    position: absolute;
    right: 2px;
    top: 4px;
    width: 20px;
    z-index: -1;
}

.image-preview {
    height: 200%;
    float: left;
}
</style>
