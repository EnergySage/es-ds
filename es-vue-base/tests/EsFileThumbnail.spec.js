import { mount } from '@vue/test-utils';
import EsFileThumbnail from '@/src/lib-components/EsFileThumbnail.vue';
import jestVue from '@/tests/jest.vue.config';

describe('EsFileThumbnail', () => {
    test('<EsFileThumbnail />', async () => {
        const wrapper = mount(EsFileThumbnail, {
            ...jestVue,
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    test('check for <EsFileThumbnail />', async () => {
        const wrapper = mount(EsFileThumbnail, {
            ...jestVue,
        });

        expect(wrapper.find('.mobile-preview').exists()).toBe(true);
        expect(wrapper.find('.desktop-preview').exists()).toBe(true);
    });

    // test default prop values
    test('<EsFileThumbnail /> default prop values', async () => {
        const wrapper = mount(EsFileThumbnail, {
            ...jestVue,
        });
        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();

        expect(wrapper.props('percentLoaded')).toBe(0);
        expect(wrapper.props('previewText')).toBe('View preview');
    });
});
