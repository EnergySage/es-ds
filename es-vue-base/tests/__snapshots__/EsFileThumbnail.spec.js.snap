// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsFileThumbnail <EsFileThumbnail /> 1`] = `
"<div>
  <div class="mobile-preview d-block d-md-none">
    <div class="card thumbnail-border">
      <div class="row p-3 align-items-center">
        <div class="col-2">
          <!---->
          <!---->
        </div>
        <div class="p-0 col-8">
          <div class="font-weight-bolder text-gray-800 pl-0">
            test.png
          </div>
        </div>
        <div class="col-2"><a aria-label="close-file-mobile" href="#" target="_self" class="text-gray-900 float-right icon-button"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="currentColor" style="height: 24px; width: 24px;">
              <path d="M18.62 5.081h3.384c.598 0 .996.399.996.996s-.398.995-.995.995H20.91l-1.494 13.24c-.1.697-.398 1.394-.995 1.892-.598.497-1.294.796-2.09.796H7.57c-.698 0-1.395-.299-1.992-.796-.597-.498-.896-1.195-.995-1.892L3.09 7.072H1.994c-.597 0-.995-.398-.995-.995s.398-.996.995-.996H5.38l1.095-2.29c.2-.497.597-.995 1.095-1.293C8.068 1.199 8.566 1 9.163 1h5.674c.597 0 1.095.2 1.593.498s.896.697 1.095 1.294l1.095 2.29Z" fill="none"></path>
              <path d="M22.005 5.081H18.62l-1.095-2.29c-.2-.596-.597-.995-1.095-1.293-.498-.299-.996-.498-1.593-.498H9.163c-.597 0-1.095.2-1.593.498a2.752 2.752 0 0 0-1.095 1.294L5.38 5.082H1.995c-.597 0-.995.398-.995.995s.398.995.995.995H3.09l1.494 13.24c.1.697.398 1.394.995 1.892.597.497 1.294.796 1.991.796h8.76c.797 0 1.494-.299 2.09-.796.598-.498.897-1.195.996-1.892l1.494-13.24h1.095c.597 0 .995-.398.995-.995s-.398-.996-.995-.996ZM8.267 3.588c0-.1.2-.298.398-.298.1-.1.299-.2.498-.2h5.575c.199 0 .298.1.497.2.2.1.299.199.399.398l.696 1.393H7.57l.697-1.493Zm9.158 16.525c0 .2-.1.398-.298.597-.2.2-.498.3-.697.3H7.57c-.299 0-.498-.1-.697-.3-.199-.1-.298-.398-.298-.597L5.08 7.073h13.84l-1.494 13.04Z"></path>
              <path d="M13.941 17.027a.998.998 0 0 0 .996-.995v-4.978c0-.597-.398-.995-.996-.995-.597 0-.995.398-.995.995v4.978c0 .597.398.995.995.995Zm-3.981 0a.997.997 0 0 0 .995-.995v-4.978c0-.597-.398-.995-.996-.995-.597 0-.995.398-.995.995v4.978c0 .597.398.995.995.995Z"></path>
            </svg></a></div>
      </div>
      <div class="row px-3 pb-3 align-items-center">
        <div class="col-2"></div>
        <div class="pl-0 col-10">
          <!---->
        </div>
      </div>
    </div>
  </div>
  <div class="desktop-preview d-none d-md-block">
    <div class="thumbnail-outer-wrapper">
      <div class="float-right d-flex">
        <!---->
        <div class="pt-2"></div>
      </div> <a aria-label="show-preview-desktop" href="#" target="_self" tabindex="-1" aria-disabled="true" class="text-decoration-none text-gray-800 disabled">
        <div class="card thumbnail-inner-wrapper">
          <div class="h-100 d-flex align-items-center justify-content-center">
            <div><svg viewBox="0 0 100 100" height="65px" width="65px">
                <circle cx="50" cy="50" stroke-width="5" r="45%" class="inner-circle"></circle>
                <!----> <text x="50%" y="52%" dominant-baseline="middle" alignment-baseline="middle" text-anchor="middle" class="font-weight-boldest text-gray-700 font-size-base">
                  0%
                </text>
              </svg></div>
          </div>
        </div>
      </a>
    </div>
    <div class="font-weight-bolder pt-2 pt-md-3">
      test.png
    </div>
    <!---->
  </div>
</div>"
`;

exports[`EsFileThumbnail <EsFileThumbnail /> default prop values 1`] = `
"<div>
  <div class="mobile-preview d-block d-md-none">
    <div class="card thumbnail-border">
      <div class="row p-3 align-items-center">
        <div class="col-2">
          <!---->
          <!---->
        </div>
        <div class="p-0 col-8">
          <div class="font-weight-bolder text-gray-800 pl-0">
            test.png
          </div>
        </div>
        <div class="col-2"><a aria-label="close-file-mobile" href="#" target="_self" class="text-gray-900 float-right icon-button"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="currentColor" style="height: 24px; width: 24px;">
              <path d="M18.62 5.081h3.384c.598 0 .996.399.996.996s-.398.995-.995.995H20.91l-1.494 13.24c-.1.697-.398 1.394-.995 1.892-.598.497-1.294.796-2.09.796H7.57c-.698 0-1.395-.299-1.992-.796-.597-.498-.896-1.195-.995-1.892L3.09 7.072H1.994c-.597 0-.995-.398-.995-.995s.398-.996.995-.996H5.38l1.095-2.29c.2-.497.597-.995 1.095-1.293C8.068 1.199 8.566 1 9.163 1h5.674c.597 0 1.095.2 1.593.498s.896.697 1.095 1.294l1.095 2.29Z" fill="none"></path>
              <path d="M22.005 5.081H18.62l-1.095-2.29c-.2-.596-.597-.995-1.095-1.293-.498-.299-.996-.498-1.593-.498H9.163c-.597 0-1.095.2-1.593.498a2.752 2.752 0 0 0-1.095 1.294L5.38 5.082H1.995c-.597 0-.995.398-.995.995s.398.995.995.995H3.09l1.494 13.24c.1.697.398 1.394.995 1.892.597.497 1.294.796 1.991.796h8.76c.797 0 1.494-.299 2.09-.796.598-.498.897-1.195.996-1.892l1.494-13.24h1.095c.597 0 .995-.398.995-.995s-.398-.996-.995-.996ZM8.267 3.588c0-.1.2-.298.398-.298.1-.1.299-.2.498-.2h5.575c.199 0 .298.1.497.2.2.1.299.199.399.398l.696 1.393H7.57l.697-1.493Zm9.158 16.525c0 .2-.1.398-.298.597-.2.2-.498.3-.697.3H7.57c-.299 0-.498-.1-.697-.3-.199-.1-.298-.398-.298-.597L5.08 7.073h13.84l-1.494 13.04Z"></path>
              <path d="M13.941 17.027a.998.998 0 0 0 .996-.995v-4.978c0-.597-.398-.995-.996-.995-.597 0-.995.398-.995.995v4.978c0 .597.398.995.995.995Zm-3.981 0a.997.997 0 0 0 .995-.995v-4.978c0-.597-.398-.995-.996-.995-.597 0-.995.398-.995.995v4.978c0 .597.398.995.995.995Z"></path>
            </svg></a></div>
      </div>
      <div class="row px-3 pb-3 align-items-center">
        <div class="col-2"></div>
        <div class="pl-0 col-10">
          <!---->
        </div>
      </div>
    </div>
  </div>
  <div class="desktop-preview d-none d-md-block">
    <div class="thumbnail-outer-wrapper">
      <div class="float-right d-flex">
        <!---->
        <div class="pt-2"></div>
      </div> <a aria-label="show-preview-desktop" href="#" target="_self" tabindex="-1" aria-disabled="true" class="text-decoration-none text-gray-800 disabled">
        <div class="card thumbnail-inner-wrapper">
          <div class="h-100 d-flex align-items-center justify-content-center">
            <div><svg viewBox="0 0 100 100" height="65px" width="65px">
                <circle cx="50" cy="50" stroke-width="5" r="45%" class="inner-circle"></circle>
                <!----> <text x="50%" y="52%" dominant-baseline="middle" alignment-baseline="middle" text-anchor="middle" class="font-weight-boldest text-gray-700 font-size-base">
                  0%
                </text>
              </svg></div>
          </div>
        </div>
      </a>
    </div>
    <div class="font-weight-bolder pt-2 pt-md-3">
      test.png
    </div>
    <!---->
  </div>
</div>"
`;
